### PR TITLE
gcc 8.4 updates

### PR DESCRIPTION
**Pull request by ltirrell**
## Overview
We are building the latest FreeSurfer dev in an Ubuntu 18 (bionic) Docker container, using its default `gcc-8` version (8.4). The container has requirements installed based off of the [BuildRequirements](https://surfer.nmr.mgh.harvard.edu/fswiki/BuildRequirements) page. Exact versions and flags are listed below (we're using the minimal build with no GUIs).  **NOTE:** I ran `make test`, which had 6 failures. I'm not sure if any of these are expected. More info is below, and I am debugging now.

## Summary of changes
I'm mostly copy-pasting from Richards previous PRs (#784  and #834, among others):
    - changing sprintf() to snprintf()
    - adding ignores for `fallthrough`/`alloc-size-larger-than`
    - Fixes for `class-memaccess`

Additionally appears to be a `stringop-overflow` bug in `DICOMRead.cpp`, which as a proposed fix.

With our current build options, we needed to `find_library` for `libquadmath` in order to build `mri_robust_register`, which was added to the top-level CMakeLists. I could move this to the specific sub-dir if that is more appropriate.

`mris_exvivo_surfaces` was removed from the MINIMAL build, since it doesn't seem necessary for recon-all (and I wanted to reduce the amount of fixes I needed to make)

`cmake` doesn't like the `-j8` flag used when building packages. This was globally replaced with `-j 8`

more building, testing, and binary files were added to the gitignore

`mri_cal_normalize` was moved to the attic

### Versions
- cmake version 3.12.0
- gcc (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0
- g++ (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0
- GNU Fortran (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0

### `cmake` flags:
    -DBUILD_GUIS=OFF 
    -DMINIMAL=ON 
    -DQATOOLS_MODULE=OFF 
    -DCMAKE_BUILD_TYPE=Release
    -DINSTALL_PYTHON_DEPENDENCIES=OFF
    -DDISTRIBUTE_FSPYTHON=OFF
    -DCMAKE_INSTALL_PREFIX="${PREFIX_DIR}"
    -DFS_PACKAGES_DIR="${FREESURFER_CODE_DIR}/packages"

### `build_packages.py` command
```
./build_packages.py --only-itk  "${FREESUFER_CODE_DIR}/packages"
```
### Failed tests
- 22:mri_normalize_test
- 44:mris_ca_label_test
- 49:mris_curvature_stats_test
- 54:mris_fix_topology_test
- 60:mris_make_surfaces_test
- 61:mris_place_surface_test

### Log

[LastTest.log](https://github.com/freesurfer/freesurfer/files/5882678/LastTest.log)

There are also many instances of this:
```
error: No such file or directory
error: Could not set locale
```
which I believe doesn't effect test results, and is most likely due to the way our container is configured,

### CODE REVIEW

**Comment by ltirrell** [comment]
Here is a summary of the failed tests (from the above attached log). Tests 22, 54, 60 and 61 failed due to `mri[s]_diff` of the results, and 49 failed due to `diff`. These may just be minor differences due to running on a separate platform from where references were created (the diff errors in 49 are in the 7th sig fig)?

However, in test 44  `mris_ca_label` failed to completion. I don't think I changed anything related to this so unsure of why this failed

- 22:mri_normalize_test
  - for `mri_diff brain.mgz brain.ref.mgz --debug`:
    - maxdiff 13.000000 at 93 106 115 0
    - diffcount 117841
    - Volumes differ in pixel data
    - maxdiff  13.00000000 at 93 106 115 0
- 44:mris_ca_label_test
  - for `mris_ca_label -l bert/label/lh.cortex.label -aseg bert/mri/aseg.mgz -seed 1234 bert lh bert/surf/lh.sphere.reg /code/distribution/average/lh.curvature.buckner40.filled.desikan_killiany.2010-03-25.gcs bert/label/lh.aparc.annot`:
    - error: Cannot allocate memory
    - error: read_ico: could not allocate vertex list
- 49:mris_curvature_stats_test
  - difference in `diff raw ref` (last few decmial places are off)
- 54:mris_fix_topology_test
  - for `mris_diff subj1/surf/lh.orig subj1/surf/lh.orig.ref --debug`:
    - Number of vertices 152773 152780
    - Number of faces    305542 305556
    - Surfaces differ in number of vertices 152773 152780 Consider using --renumbered
- 60:mris_make_surfaces_test
  - for `mris_diff subject/surf/lh.pial subject/surf/lh.pial.REF --debug`:
    - Too many differences in face nxnynz (and maybe others)
- 61:mris_place_surface_test
    - for `mris_diff subject/surf/lh.pial subject/surf/lh.pial.REF --debug`:
      - Too many differences in face nxnynz (and maybe others)

File: mri_watershed/brain_volume/mri_brain_volume.cpp [hunk 1]
```diff
@@ -1044,7 +1047,14 @@ static int Pre_CharSorting(STRIP_PARMS *parms,MRI_variables *MRI_var) {
...
     for (j=0;j<r;j++)
       for (i=0;i<r;i++) {
+#if GCC_VERSION > 80000
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wint-in-bool-context"
+#endif
         if (!(i*j*k*(i-r+1)*(j-r+1)*(k-r+1))) {
+#if GCC_VERSION > 80000
+#pragma GCC diagnostic pop
+#endif
```
Comment (ltirrell) [inline mri_watershed/brain_volume/mri_brain_volume.cpp]: note that the error for this (and `mri_watershed`) said:
```
/code/mri_watershed/brain_volume/mri_brain_volume.cpp:1051:36: error: '*' in boolean context, suggest '&&' instead [-Werror=int-in-bool-context]
```
(pointing to the last `*` in line 1054). I'm not sure what that means/what's happening right here so I ignored the warning)

**Comment by ltirrell** [comment]
after latest changes, "mri_watershed_test" and "talairach_avi_test" also fail with make test. full log is here:

[LastTest.log](https://github.com/freesurfer/freesurfer/files/5884702/LastTest.log)

File: mri_watershed/brain_volume/mri_brain_volume.cpp (hunk 1 above)
Comment (ahoopes) [reply mri_watershed/brain_volume/mri_brain_volume.cpp]: My guess is that something like `((i*j*k*(i-r+1)*(j-r+1)*(k-r+1)) == 0)` would silence this

Comment (ltirrell) [reply mri_watershed/brain_volume/mri_brain_volume.cpp]: ok i made this this change in the 2 places where it occurs

**Comment by ahoopes** [comment]
Thanks Lee!

**Comment by WangzLiang** [comment]
Hello,
     I am creating my own Gaussian Classifier Surface Arrays file using the command:mris_ca_label.I want to check the information in the .gcs file, could you provide some methods for reading the kind of files?

**Pull request merged**